File: solutions/python/queen-attack/1/queen_attack.py
class Queen:
    def __init__(self, row, col):
        if row < 0:
            raise ValueError("row not positive")
        if row > 7:
            raise ValueError("row not on board")
        if col < 0:
            raise ValueError("column not positive")
        if col > 7:
            raise ValueError("column not on board")
        self.row = row
        self.col = col

    def can_attack(self, other):
        if self.row == other.row and self.col == other.col:
            raise ValueError(
                "Invalid queen position: both queens in the same square"
            )
        return (
            self.row == other.row
            or self.col == other.col
            or abs(self.row - other.row) == abs(self.col - other.col)
        )

File: solutions/javascript/queen-attack/1/queen-attack.js
const SIZE = 8;

const validate = (row, col) => {
  if (row < 0 || row >= SIZE || col < 0 || col >= SIZE) {
    throw new Error('Queen must be placed on the board');
  }
};

export class QueenAttack {
  constructor({
    black: [blackRow, blackColumn] = [0, 3],
    white: [whiteRow, whiteColumn] = [7, 3],
  } = {}) {
    validate(whiteRow, whiteColumn);
    validate(blackRow, blackColumn);
    if (whiteRow === blackRow && whiteColumn === blackColumn) {
      throw new Error('Queens cannot share the same space');
    }
    this._white = [whiteRow, whiteColumn];
    this._black = [blackRow, blackColumn];
  }

  get white() {
    return this._white;
  }

  get black() {
    return this._black;
  }

  get canAttack() {
    const [wr, wc] = this._white;
    const [br, bc] = this._black;
    return wr === br || wc === bc || Math.abs(wr - br) === Math.abs(wc - bc);
  }

  toString() {
    const board = Array.from({ length: SIZE }, () => new Array(SIZE).fill('_'));
    const [wr, wc] = this._white;
    const [br, bc] = this._black;
    board[wr][wc] = 'W';
    board[br][bc] = 'B';
    return board.map((row) => row.join(' ')).join('\n');
  }
}
